### PR TITLE
Introduce VoteState::deserialize_into_uninit

### PR DESCRIPTION
**Pull request by alessandrod**
Deserializing into `MaybeUninit<VoteState>` saves the extra cost of initializing into a value initialized with VoteState::default().

I am planning to use this to speed up VoteAccount::try_from() in a followup PR.

### CODE REVIEW

**Comment by alessandrod** [comment]
self reminder: add SAFETY markers to all unsafe blocks

**Comment by alessandrod** [comment]
This PR breaks the API (the breakage is warranted IMO), and now SPL tests are failing. But this is chicken and egg? How do I land this PR if the SPL tests fail, but how do I fix the SPL tests for the new API if I can't land this? 

Maybe I could add a new `VoteState::deserialize_into_uninit`, land, deprecate `VoteState::deserialize_into` then rename `deserialize_into_uninit` to `deserialize_into`?

**Comment by alessandrod** [comment]
> Maybe I could add a new `VoteState::deserialize_into_uninit`

I've done this now

**Comment by 2501babe** [comment]
had to read the docs on `MaybeUninit<T>` and `addr_of_mut`, am i correct in thinking:
* the `unsafe` writes are safe because the typechecker ensures anything passed as `vote_state: &mut MaybeUninit<VoteState>,` is properly sized (ie protected from out of bounds writes)?
* furthermore all `unsafe` writes are to the `Sized` fields of the `struct` only. eg you allocate the dynamic vectors normally and merely write the pointer/capacity/length tripe into `VoteState`?
* because all the addresses are given to `write` via field references, it doesnt matter that the structs are `repr(Rust)`?

just want to be sure i understand everything right since `deserialize_into` is meant to be used in bpf programs, but seems like this is a "safe as long as the caller behaves" and doesnt depend on the input buffer at all

**Comment by alessandrod** [comment]
> had to read the docs on `MaybeUninit<T>` and `addr_of_mut`, am i correct in thinking:
> 
>     * the `unsafe` writes are safe because the typechecker ensures anything passed as `vote_state: &mut MaybeUninit<VoteState>,` is properly sized (ie protected from out of bounds writes)?

Correct! Both in the `deserialize_into()` and `deserialize_into_uninit()` cases the type checker checks that the memory layout is correct; since we call `deserialize_into_ptr` from those, we can guarantee that the pointer passed is always correct.


>     * furthermore all `unsafe` writes are to the `Sized` fields of the `struct` only. eg you allocate the dynamic vectors normally and merely write the pointer/capacity/length tripe into `VoteState`?

Correct.

> 
>     * because all the addresses are given to `write` via field references, it doesnt matter that the structs are `repr(Rust)`?

`write()` doesn't care about type layout, it only cares about the type. So if you have `let ptr: *mut T` you can always do `let v = ..somehow create a value of type T...; ptr.write(v)` as long as  `ptr` is a valid pointer. What makes `write()` unique is that it doesn't drop the previous value, eg consider:

```
#[derive(Debug)]
struct A {
    msg: String
}

fn main() {
    let mut a: std::mem::MaybeUninit<A> = std::mem::MaybeUninit::uninit();
    let a = unsafe {
        let a_ptr = a.as_mut_ptr();
        // this is undefined behaviour and likely segfaults, since it drops the previous `msg` but it hasn't been allocated yet so it points to a random address
        // *std::ptr::addr_of_mut!((*a_ptr).msg) = String::from("hello");
        // this is correct
        std::ptr::addr_of_mut!((*a_ptr).msg).write(String::from("hello"));
        a.assume_init()
    };
    eprintln!("{a:?}")
}
```

> just want to be sure i understand everything right since `deserialize_into` is meant to be used in bpf programs, but seems like this is a "safe as long as the caller behaves" and doesnt depend on the input buffer at all

Yes it's meant to _always_ be safe as long as you use safe rust, independent from what's inside the buffer.

**Comment by alessandrod** [comment]
> just want to be sure i understand everything right since deserialize_into is meant to be used in bpf programs, but seems like this is a "safe as long as the caller behaves" and doesnt depend on the input buffer at all

also mind that I haven't added a test, but this fixes deserialization for the following case:

```
let mut vote_state = VoteState::default();
VoteState::deserialize_into(data, &mut vote_state).unwrap();
VoteState::deserialize_into(data, &mut vote_state).unwrap();
```

In master this will incorrectly _append_ votes, authorized_voters etc twice. With this PR the 2nd `deserialize_into` resets the passed `&mut VoteState` before deserializing, so fields don't get "duplicated".

**Comment by alessandrod** [comment]
@ryoqun just tagged you too for review to double check the unsafe bits if you can

File: sdk/program/src/vote/state/mod.rs
Comment (brooksprumo) [inline sdk/program/src/vote/state/mod.rs:557]: Why have this fn take a MaybeUninit parameter and force the caller to correctly set it initialized? We have all that information here, yes?

Have the return value be a `Result<VoteState, InstructionError>`

And then the impl would be:

```rust
let mut vote_state = MaybeUninit::uninit();
VoteState::deserialize_into_ptr(input, vote_state.as_mut_ptr())?;
let vote_state = unsafe { vote_state.assume_init() };
Ok(vote_state)
```

Maybe the issue is because other callers cannot do this pattern?

**Review by brooksprumo — COMMENTED** [review]
I'm still reviewing the rest, this was the first thing that caught my eye.

File: sdk/program/src/vote/state/mod.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs:557]: Good question!

The point of this function is to deserialize into a _provided_ `VoteState` and not create a new `VoteState`. What you are suggesting always creates a new value (`let mut vote_state = MaybeUninit::uninit()`). 

There are two reasons for wanting to deserialize into a provided value:

- take `VoteAccount`, where `VoteState` is a subfield of `VoteAccount` - by deserializing into a given `VoteState` I can deserialize in place, skipping a memcpy.
- what you're suggesting is similar to what `serde::Deserialize::deserialize` does. Because LLVM/rustc aren't great at NRVO, even if you do `let state = VoteState::deserialize(input)` often the compiler won't create the result straight in the caller frame, but it will create it in the callee frame and then memcpy it back to the caller. In a followup PR, I show how this is currently happening when we do `Arc::new(vote_account)` inside `VoteAccount::try_from(data)`, where rustc first creates `vote_account` then memcpys it into the `Arc` allocation because computers were a mistake and everything is terrible

Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs:557]: >  will create it in the callee frame and then memcpy it back to the caller

should mention that this is especially bad for `VoteState` because `VoteState::prior_voters` is a 1k byte array so memcpy cost is not negligible

Comment (brooksprumo) [reply sdk/program/src/vote/state/mod.rs:557]: > What you are suggesting always creates a new value (`let mut vote_state = MaybeUninit::uninit()`).

Yep!

> take `VoteAccount`, where `VoteState` is a subfield of `VoteAccount` - by deserializing into a given `VoteState` I can deserialize in place, skipping a memcpy.

Ahh, that's the context I didn't have. So the real caller will have an uninitialized VoteAccount, and then call `deserialize_into_uninit()`? And that would be to avoid having to call `deserialize_into_ptr()` directly then?

> Because LLVM/rustc aren't great at NRVO, even if you do `let state = VoteState::deserialize(input)` often the compiler won't create the result straight in the caller frame

😭 

> because computers were a mistake and everything is terrible

😂

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (brooksprumo) [inline sdk/program/src/vote/state/vote_state_deserialize.rs]: Do you think it would also be useful to mention that we *must* use `addr_of_mut!()` here and not "normal" casting?

Comment (brooksprumo) [inline sdk/program/src/vote/state/vote_state_deserialize.rs]: And how much would you dislike moving the safe bits out of the unsafe block?

```suggestion
    let node_pubkey = read_pubkey(cursor)?;
    let authorized_withdrawer = read_pubkey(cursor)?;
    let commission = read_u8(cursor)?;
    unsafe {
        addr_of_mut!((*vote_state).node_pubkey).write(node_pubkey);
        addr_of_mut!((*vote_state).authorized_withdrawer).write(authorized_withdrawer);
        addr_of_mut!((*vote_state).commission).write(commission);
    }
```

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (brooksprumo) [inline sdk/program/src/vote/state/vote_state_deserialize.rs]: nit? Maybe this makes the `read_u64()` calls later less clear?
```suggestion
        let prior_voters_buf = addr_of_mut!((*prior_voters).buf) as *mut (Pubkey, Epoch, Epoch);
```

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs]: yep good point!

Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs]: Agree this is better

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs]: Nope I think I was just lazy, good point

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs]: done

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs]: done

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs]: done

File: sdk/program/src/vote/state/mod.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs:557]: > Ahh, that's the context I didn't have. So the real caller will have an uninitialized VoteAccount, and then call deserialize_into_uninit()? And that would be to avoid having to call deserialize_into_ptr() directly then?

correct! VoteAccount is in another crate. I don't want to expose `deserialize_into_ptr` and pointers are gross anyway, `MaybeUninit` is what rust provides to do what we want to do here.

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (brooksprumo) [inline sdk/program/src/vote/state/vote_state_deserialize.rs]: For the impl of this whole function, is it better to read-then-write fields one at a time, or would it be better to read all the fields, then write all the fields? Mostly I'm not sure I understand why we do something in the middle.

Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs]: No reason!

I've pushed a commit that moves _some_ writes at the bottom, but I've also changed the code to read pubkeys straight into their destination avoiding a memcpy

File: sdk/program/src/serialize_utils/cursor.rs
Comment (brooksprumo) [inline sdk/program/src/serialize_utils/cursor.rs]: Why call `fill_buf` again here? Can we use `buf` directly, since we've already got it, and ensured its size is correct?

```suggestion
            // Safety: `buf` is guaranteed to be at least `PUBKEY_SIZE` bytes long
            unsafe {
                ptr::copy_nonoverlapping(buf.as_ptr(), pubkey as *mut u8, PUBKEY_SIZE);
            }

            cursor.consume(PUBKEY_SIZE);
```

Comment (alessandrod) [reply sdk/program/src/serialize_utils/cursor.rs]: err 🤦‍♂️ I first wrote the let else, then had to check the length so I figured I'd just use a match statement... but forgot to remove the earlier code. I actually prefer the match statement

Comment (alessandrod) [reply sdk/program/src/serialize_utils/cursor.rs]: fixed!

File: sdk/program/src/vote/state/mod.rs
Comment (ryoqun) [inline sdk/program/src/vote/state/mod.rs]: nit: `Self::deserialzie_into_ptr()`? (i.e. can use `Self`).

File: sdk/program/src/vote/state/mod.rs
Comment (ryoqun) [inline sdk/program/src/vote/state/mod.rs:556]: nit: `Self::deserialzie_into_ptr()`? (i.e. can use `Self`).

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (ryoqun) [inline sdk/program/src/vote/state/vote_state_deserialize.rs:55]: hmm, i can assume this deserialization can be run on untrusted data, right? (i.e. it's fallible as coded with `?`s).

in that case, i think these fields need to be initialized in the case of errors as well (i.e. when bailing out with one of those above `?`s hit by `InvalidAccountData`).

Otherwise, `std::ptr::drop_in_place()`-ed `vote_state: &mut VoteState` will leave some fields uninitialzied when returning `Err(_)` from `deserialize_into()`. And it will cause uaf&double-free eventually when the referenced `VoteState` is dropped sometime later, right?

On the other hand, `deserialize_into_uninit(vote_state: &mut MaybeUninit<VoteState>)` should be safe, assuming `assume_init()` won't called on `vote_state` at the caller when `Err(_)` is returned from this `fn`.

File: sdk/program/src/serialize_utils/cursor.rs
Comment (ryoqun) [inline sdk/program/src/serialize_utils/cursor.rs]: nit: how about commenting that this safety also rely on the fact `align = 1` for `Pubkey`?

**Comment by ryoqun** [comment]
fyi, it looks like miri passes: f9e974ff5c4d4fe450e5fcc821a9cd56577e2926 (need to severely reduce the loop counts due to miri being too slow.. lol)

```
$ cargo miri test -p solana-program vote -- --skip compact
Preparing a sysroot for Miri (target: x86_64-unknown-linux-gnu)... done
   Compiling solana-program v2.1.0 (/home/ryoqun/work/solana/for-small-patches2/sdk/program)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.47s
     Running unittests src/lib.rs (target/miri/x86_64-unknown-linux-gnu/debug/deps/solana_program-e0bef1dfcfb6cd11)

running 22 tests
test vote::error::tests::test_custom_error_decode ... ok
test vote::program::test_id ... ok
test vote::state::tests::test_authorized_voter_is_locked_within_epoch ... ok
test vote::state::tests::test_circbuf_oob ... ok
test vote::state::tests::test_default_vote_state_is_uninitialized ... ok
test vote::state::tests::test_get_and_update_authorized_voter ... ok
test vote::state::tests::test_is_correct_size_and_initialized ... ok
test vote::state::tests::test_minimum_balance ... ok
test vote::state::tests::test_set_new_authorized_voter ... ok
test vote::state::tests::test_vote_deserialize_into_uninit ... ok
test vote::state::tests::test_vote_deserialize_into_uninit_ill_sized ... ok
test vote::state::tests::test_vote_deserialize_into_uninit_nopanic ... ok
test vote::state::tests::test_vote_process_timestamp ... ok
test vote::state::tests::test_vote_serialize ... ok
test vote::state::tests::test_vote_state_commission_split ... ok
test vote::state::tests::test_vote_state_epoch0_no_credits ... ok
test vote::state::tests::test_vote_state_epoch_credits ... ok
test vote::state::tests::test_vote_state_increment_credits ... ok
test vote::state::tests::test_vote_state_max_size ... ok
test vote::state::tests::test_vote_state_size_of ... ok
test vote::state::vote_state_0_23_5::tests::test_vote_deserialize_0_23_5 ... ok
test vote::state::vote_state_1_14_11::tests::test_vote_deserialize_1_14_11 ... ok

test result: ok. 22 passed; 0 failed; 0 ignored; 0 measured; 232 filtered out; finished in 164.07s
```

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (ryoqun) [reply sdk/program/src/vote/state/vote_state_deserialize.rs:55]: considering `deserialize_into` impl will be deprecated and short-lived, how about just restoring old impl instead of flowing into the shiny `deserialize_into_uninit()`?

Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs:55]: we'd have to duplicate all the code, since I moved the old code to working with `*mut VoteState`. I think there's a better fix, let me do that and add a test.

File: sdk/program/src/vote/state/mod.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs:556]: There are other pre-existing bits that already use VoteState instead of Self, so I'll leave it to VoteState (which also happens to be my preference :P)

File: sdk/program/src/serialize_utils/cursor.rs
Comment (alessandrod) [reply sdk/program/src/serialize_utils/cursor.rs]: done

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs:55]: I've pushed a fix for this. When deserialize_into_ptr fails, we reset to `VoteState::default()`

File: sdk/program/src/vote/state/mod.rs
Comment (ryoqun) [inline sdk/program/src/vote/state/mod.rs:502]: nit: i like this recently stabilized one over the good old `as *mut T` casts as documented there and for better readability (imo): https://doc.rust-lang.org/std/ptr/fn.from_mut.html

File: sdk/program/src/vote/state/mod.rs
Comment (ryoqun) [inline sdk/program/src/vote/state/mod.rs]: maybe you intentionally avoided `.map_err()`?. how about the new guy called `inspect_err()` for conciseness (again imo)?: https://doc.rust-lang.org/std/result/enum.Result.html#method.inspect_err

File: sdk/program/src/vote/state/mod.rs
Comment (brooksprumo) [reply sdk/program/src/vote/state/mod.rs:502]: +1

I'm a fan of `ptr::from_mut(var).cast()` too (vs `var as *mut _ as *mut u8`)

File: sdk/program/src/vote/state/mod.rs
Comment (brooksprumo) [reply sdk/program/src/vote/state/mod.rs]: I think alessandro is a match fan. I prefer *not* using a `.map_err()`/`.inspect_err()` when there are side effects (writing the default VoteState in this case). If it were me, I'd do:

```rust
let result = VoteState::deserialize_into_ptr(..);
if result.is_err() {
    unsafe { vote_state.write(VoteState::default()); }
}
result
```

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (ryoqun) [inline sdk/program/src/vote/state/vote_state_deserialize.rs]: nit: the first `u64` is `Epoch`.

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (ryoqun) [inline sdk/program/src/vote/state/vote_state_deserialize.rs:146]: nit: `// Safety: ...` is missing

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (ryoqun) [inline sdk/program/src/vote/state/vote_state_deserialize.rs:102]: nit: this `as *mut (...)` can be replaced with `slice`'s `as_mut_ptr()`, which is less error-prone, imo.

File: sdk/program/src/serialize_utils/cursor.rs
Comment (ryoqun) [inline sdk/program/src/serialize_utils/cursor.rs]: nit: strictly speaking, i think `solana_program::pubkey::PUBKEY_BYTES` is more appropriate than this because this code is dealing with reading N bytes from the buffer. however, `mem::size_of::<Pubkey> == PUBKEY_BYTES` (no padding or whatsoever). and i don't see any possibility of diverting the equality in foreseeable future. so quite minor thing ;)

File: sdk/program/src/vote/state/mod.rs
Comment (ryoqun) [inline sdk/program/src/vote/state/mod.rs]: this looks okay.

nit: however, for completeness, similar cleaning is desired when panicking. that said, i couldn't think of panic arising from `VoteState::deserialize_into_ptr()`. so, just mentioning about no panic will be enough.

File: sdk/program/src/vote/state/mod.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs]: imagine ever using is_err!  And yeah I don't like using map_* or inspect when there are side effects.

In cases like this, I am indeed a match fan, it feels more idiomatic than C style err = ...; if err != 0 { ... }

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs:146]: oops!

**Comment by ryoqun** [comment]
> `deserialize_into` is meant to be used in bpf programs

just curious but i wonder how the CUs is reduced after this patch to save kittens. :)

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs:102]: I don't think so? that would create a reference (&self to as_mut_ptr()) which is UB

**Comment by alessandrod** [comment]
> > `deserialize_into` is meant to be used in bpf programs
> 
> just curious but i wonder how the CUs is reduced after this patch to save kittens. :)

that'd be interesting, but it's not why I'm doing this. I'm doing this to fix `VoteAccount::try_from`:

```
before: test bench_vote_account ... bench:       5,214 ns/iter (+/- 1,102)
after: test bench_vote_account ... bench:         326 ns/iter (+/- 85)
```

File: sdk/program/src/vote/state/mod.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs:502]: Nod, I'll just never remember that ptr::from_mut exists

File: sdk/program/src/vote/state/mod.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs]: yep nice catch, fixed the panic case too

File: sdk/program/src/serialize_utils/cursor.rs
Comment (alessandrod) [reply sdk/program/src/serialize_utils/cursor.rs]: done

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs:146]: done

File: sdk/program/src/vote/state/vote_state_deserialize.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/vote_state_deserialize.rs]: done

File: sdk/program/src/vote/state/mod.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs]: I actually ended up removing the match 😭

File: sdk/program/src/vote/state/mod.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs:502]: done

Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs:502]: This is giving me

```
error: current MSRV (Minimum Supported Rust Version) is `1.75.0` but this item is stable since `1.76.0`
   --> sdk/program/src/vote/state/mod.rs:497:26
    |
497 |         let vote_state = ptr::from_mut(vote_state);
    |                          ^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#incompatible_msrv
    = note: `-D clippy::incompatible-msrv` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::incompatible_msrv)]`
```

which is weird since there are already uses of it from accounts-db?

Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs:502]: Reverted since platform-tools is still on rust 1.75

**Review by brooksprumo — DISMISSED** [review]
:shipit:

File: sdk/program/src/vote/state/mod.rs
Comment (ryoqun) [inline sdk/program/src/vote/state/mod.rs]: nit: `rustc >= 1.76`

File: sdk/program/src/serialize_utils/cursor.rs
Comment (ryoqun) [inline sdk/program/src/serialize_utils/cursor.rs]: nit: ... to be at least `PUBKEY_BYTES` bytes ...

**Review by ryoqun — DISMISSED** [review]
lgtm with nits.

the bench numbers looks promising. looking forward to see a follow-up pr for actual perf gains.

File: sdk/program/src/serialize_utils/cursor.rs
Comment (alessandrod) [reply sdk/program/src/serialize_utils/cursor.rs]: done

File: sdk/program/src/vote/state/mod.rs
Comment (alessandrod) [reply sdk/program/src/vote/state/mod.rs]: done

**Review by brooksprumo — APPROVED** [review]

**Review by ryoqun — APPROVED** [review]
lgtm _with no nit_ (really thanks for addressing all the review comments)

**Comment by alessandrod** [comment]
thanks a lot for the thorough reviews!

**Pull request merged**
